### DISCUSSION TITLE
[error] No space left on device

### DISCUSSION THREAD

**Discussion by xerxesb** [community/community#25678] [Actions]
<p>Hi,</p>
<p>I’ve created an action workflow that builds our docker image for distribution. This image requires Tensorflow, CUDA, and a number of other libs so it’s reasonably weighty coming in at about 18GB.</p>
<p>During a build run, my workflow fails part way through citing “##[error]No space left on device”</p>
<p>I think I read somewhere that Github runners have about 16GB of space, so my question is<br>
a) whether there’s a way to increase that size limit?<br>
b) How do other teams manage this problem?</p>
<p>Happy to explore alternative solutions to this…Just not sure why there seems to be very little content when I search for this specific issue.</p>
<p>Thanks.</p>

**Comment by otter-computer** [comment]
<p>Hi <a class="mention" href="/xerxesb">@xerxesb</a> , welcome to the GitHub Support Community! Yeah, GitHub-hosted runners only have 14GB of free space available, and there isn’t a way to increase that I’m afraid.</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/reference/specifications-for-github-hosted-runners#supported-runners-and-hardware-resources" target="_blank" rel="noopener">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/reference/specifications-for-github-hosted-runners#supported-runners-and-hardware-resources" target="_blank" rel="noopener">Specifications for GitHub-hosted runners - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/reference/specifications-for-github-hosted-runners#supported-runners-and-hardware-resources" target="_blank" rel="noopener">//docs.github.com/en/actions/reference/specifications-for-github-hosted-runners</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>The best solution I can think of if you need more space or a more powerful machine would be to host your own runner:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners" target="_blank" rel="noopener">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners" target="_blank" rel="noopener">About self-hosted runners - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners" target="_blank" rel="noopener">//docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>You should be able to host this where you want—on a local machine, a VM in the cloud—and doing this should give you the flexibility to configure the specs you need for your workflows.</p>

**Comment by solodon4** [comment, marked as answer]
<p>You don’t need to go hosted runner approach. You can free enough space for a docker image by using this script from Apache Flink: <a href="https://github.com/apache/flink/blob/02d30ace69dc18555a5085eccf70ee884e73a16e/tools/azure-pipelines/free_disk_space.sh" rel="noopener nofollow ugc">free_disk_space.sh</a>.</p>
<p>I am starting with 48GB available after running it, just make sure to also add monodoc-http before mono-devel in it as the script is probably a bit outdated.</p>
<p>The bigger problem I see is that when the error in subj does happen, no logs are available for the step that failed so there is no way to investigate why you actually ran over allocated space. GitHub should at least preserve the on screen logs of the step as I regularly output data there to let me figure what exactly eat up all the space, but without that log, I’m just poking at lines of code in the dark</p>

**Reply by dnut** [reply]
While this general approach could work in theory, the script you've linked is not useful for github workflow runners, since it's clearly written for a different system. It removes less than 1 GB, and most of the things it attempts to remove are not even on the runner. This script needs significant modification to be useful. I instead solved this by just removing a single folder, `/opt/hostedtoolcache`. You can find a detailed explanation in my other post in the current thread: [#discussioncomment-5242449](#discussioncomment-5242449)

**Reply by shw1n** [reply]
Thank you @dnut -- I just added `rm -rf /opt/hostedtoolcache` and my problem was solved

This works!

**Comment by xerxesb** [comment]
<p>That’s fantastic! I ran it on our repo and free space went from 19% to 53% (I also removed some other packages in the top 100 that weren’t required for our use)</p>
<p>Thanks. This has definitely helped!</p>

**Comment by ChesterZheng** [comment]
<p>Where could I run the shell <a href="https://github.com/apache/flink/blob/02d30ace69dc18555a5085eccf70ee884e73a16e/tools/azure-pipelines/free_disk_space.sh" rel="noopener nofollow ugc">free_disk_space.sh</a>?</p>

**Comment by xerxesb** [comment]
<p>You run it in your workflow. I have it defined as one of the first steps in a job</p>

**Comment by dnut** [comment]
I freed 10 GB by deleting `/opt/hostedtoolcache`. Nothing in this folder is necessary if your workflow runs in docker. I also had to switch from using the workflow "container" configuration, to instead configuring the individual steps to run inside the container with the poorly documented `uses: docker://` syntax, because the step that deletes `/opt/hostedtoolcache` needs to run outside of the container.

```yaml
  foo:
    runs-on: ubuntu-latest
    steps:
      - name: Delete huge unnecessary tools folder
        run: rm -rf /opt/hostedtoolcache

      - uses: actions/checkout@v3

      - name: Run foo
        uses: docker://foo-image
        with: 
          args: /foo-script
```

FYI these are the contents of the folder:
```
/opt/hostedtoolcache
├── CodeQL
│   ├── 2.12.2-20230207
│   └── 2.12.3-20230217
├── Java_Temurin-Hotspot_jdk
│   ├── 11.0.18-10
│   ├── 17.0.6-10
│   └── 8.0.362-9
├── PyPy
│   ├── 3.7.13
│   ├── 3.8.16
│   └── 3.9.16
├── Python
│   ├── 3.10.10
│   ├── 3.11.2
│   ├── 3.7.15
│   ├── 3.8.16
│   └── 3.9.16
├── Ruby
│   └── 3.1.3
├── go
│   ├── 1.17.13
│   ├── 1.18.10
│   ├── 1.19.6
│   └── 1.20.1
└── node
    ├── 14.21.3
    ├── 16.19.1
    └── 18.14.2
```

**Reply by loyal812** [reply]
It working well. Thanks. <3

**Reply by ravife** [reply]
Awesome! Worked like a charm.

**Reply by klesh** [reply]
Thank you!

**Reply by sconley-datacoves** [reply]
> I also had to switch from using the workflow "container" configuration, to instead configuring the individual steps to run inside the container with the poorly documented `uses: docker://` syntax, because the step that deletes `/opt/hostedtoolcache` needs to run outside of the container.

You don't have to do this.  You can mount the offending directory in docker and blow it away.  First, make your container section look like this:

```
    container:
      image: xxxx-whatever-image
      volumes:
        - /opt/hostedtoolcache:/delete-me
```

Then add a step at the top of your steps like this:

```
    steps:
      - name: Clean Out Deletable Files
        run: rm -rf /delete-me/*
        shell: bash
```

This seems to work for me and now I don't have to use the janky `docker://` syntax.

**Comment by glesperance** [comment]
This is awesome -- got me unstuck on building my cuda images <3

**Comment by ryanhill1** [comment]
This worked like a charm, thanks!

**Comment by aguizaro** [comment]
Thanks for this!

**Comment by josephcopenhaver** [comment]
Went with this successfully and reclaimed over 12g:

```bash
cd /opt
find . -maxdepth 1 -mindepth 1 '!' -path ./containerd '!' -path ./actionarchivecache '!' -path ./runner '!' -path ./runner-cache -exec rm -rf '{}' ';'
```

Figured I'd keep the 4 items in the list as they had potential for my purposes.

You can always spin up a tmate server in an action and hop into the runner to inspect what's going on in the land of denmark: https://github.com/josephcopenhaver/gha-debian-tmate

**Reply by glesperance** [reply]
^ This one got rid of even more unused files in our case. 🙌

**Comment by arbutt1010** [comment]
Here is what I did I have added a below step at the starting of my Job  

>    steps:
>       # Aggressive disk cleanup before doing anything heavy
>       - name: Aggressive free disk space
>         run: |
>           echo "################################"
>           echo "Disk usage BEFORE cleanup:"
>           echo "################################"
>           df -h
> 
>           echo "Removing hosted toolcache..."
>           sudo rm -rf /opt/hostedtoolcache || true
> 
>           echo "Removing .NET..."
>           sudo rm -rf /usr/share/dotnet || true
> 
>           echo "Removing Android SDK..."
>           sudo rm -rf /usr/local/lib/android || true
> 
>           echo "Removing GHC / Haskell..."
>           sudo rm -rf /opt/ghc || true
>           sudo rm -rf /usr/local/.ghcup || true
> 
>           echo "Docker prune (just in case runner had leftovers)..."
>           sudo docker container prune -f || true
>           sudo docker image prune -af || true
>           sudo docker volume prune -f || true
> 
>           echo "################################"
>           echo "Disk usage AFTER cleanup:"
>           echo "################################"
>           df -h

Below are the results initially before my cleaning step ran I have 79% used space after my cleaning step ran I got 40% used space and I got from 16GB to 44GB free space 

> ################################
> Disk usage BEFORE cleanup:
> ################################
> Filesystem      Size  Used Avail Use% Mounted on
> /dev/root        72G   56G   16G  79% /
> tmpfs           3.9G   84K  3.9G   1% /dev/shm
> tmpfs           1.6G  1.1M  1.6G   1% /run
> tmpfs           5.0M     0  5.0M   0% /run/lock
> /dev/sda16      881M   62M  758M   8% /boot
> /dev/sda15      105M  6.2M   99M   6% /boot/efi
> tmpfs           795M   12K  795M   1% /run/user/1001
> /home/runner/work/rag-bot/rag-bot
> total 0
> Removing hosted toolcache...
> Removing .NET...
> Removing Android SDK...
> Removing GHC / Haskell...
> Docker prune (just in case runner had leftovers)...
> Total reclaimed space: 0B
> Total reclaimed space: 0B
> Total reclaimed space: 0B
> ################################
> Disk usage AFTER cleanup:
> ################################
> Filesystem      Size  Used Avail Use% Mounted on
> /dev/root        72G   28G   44G  40% /
> tmpfs           3.9G   84K  3.9G   1% /dev/shm
> tmpfs           1.6G  1.1M  1.6G   1% /run
> tmpfs           5.0M     0  5.0M   0% /run/lock
> /dev/sda16      881M   62M  758M   8% /boot
> /dev/sda15      105M  6.2M   99M   6% /boot/efi
> tmpfs           795M   12K  795M   1% /run/user/1001

**Comment by liudonghua123** [comment]
I use `rm -rf /opt/hostedtoolcache` to release about 6GB disk space. See details on https://github.com/liudonghua123/qwen-image-edit/actions/runs/19698704568/job/56429441521.

<img width="583" height="524" alt="Image" src="https://github.com/user-attachments/assets/c3eb8106-644d-4e3a-8488-50494ca09ebe" />

**Comment by endersonmenezes** [comment]
In an effort to help resolve this issue and make it easier for users to use, we are maintaining the following code:
https://github.com/endersonmenezes/free-disk-space

Some users:

- https://github.com/duckdb/duckdb
- https://github.com/conda/conda-build
- https://github.com/asterinas/asterinas
- https://github.com/kumahq/kuma
- https://github.com/telegramdesktop/tdesktop
- https://github.com/langgenius/dify

Simple use:
```yaml
- name: Free Disk Space for Docker
  uses: endersonmenezes/free-disk-space@v3
  with:
    remove_android: true
    remove_dotnet: true
    remove_haskell: true
    rm_cmd: "rmz"  # Faster cleanup
```

have some variations, please read the README.
